### DISCUSSION TITLE
USE_MCP23XXX_DRV option to use MCP23008/17 I2C expander as virtual button/switch/relay

### DISCUSSION THREAD

**Discussion by Bacto** [arendst/Tasmota#18124] [Q&A]
Hi,

I have a `MCP23008` I2C GPIO expander.

It works great with MCP230xx support ([USE_MCP230xx](https://github.com/arendst/Tasmota/blob/881c4f4a854566091f6e3188149258359c21e3a2/tasmota/my_user_config.h#L612)).

But I've discovered the [USE_MCP23XXX_DRV](https://github.com/arendst/Tasmota/blob/881c4f4a854566091f6e3188149258359c21e3a2/tasmota/my_user_config.h#L616) option that "Enable MCP23xxx support as virtual switch/button/relay".

For what I understand (I'm new in Tasmota world), with it, I could have the I2C expander GPIO show as extended ESP8266 GPIO.
Unfortunately I can't make it work.

I've compiled Tasmota with `USE_MCP23XXX_DRV` and then, using the [xdrv_67_mcp23xxx.ino](https://github.com/arendst/Tasmota/blob/ccdab295e738fe53a0335719aa9c41f3be43ee5d/tasmota/tasmota_xdrv_driver/xdrv_67_mcp23xxx.ino) source file informations, I've created a template:
`{"NAME":"My board","GPIO":[0,0,0,0,640,608,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1,"CMND":"rule3 on file#mcp23x.dat do {\"NAME\":\"My expander\",\"GPIO\":[224,225,226,227,32,33,34,35]} endon"}`

But then... nothing. I can't understand how it works.
I'm totally lost and tried everything I can thing of... but nothing.

Also, I can't find any documentation or information about this.

Anybody has some experience with the `USE_MCP23XXX_DRV` option?

**Comment by barbudor** [comment]
Have you activated the template? 
If not, go and check the box in Configuration - > Configure Others
Or type `Module 0`  in the console

The template should have created a Rule3
Check this worked by typing `rule3` in the console and at the end of the response you should find
```
"Rules":"on file#mcp23x.dat do {\"NAME\":\"My expander\",\"GPIO\":[224,225,226,227,32,33,34,35]} endon"
```

If not, re-enter the rule type
```
Rule3 on file#mcp23x.dat do {"NAME":"My expander","GPIO":[224,225,226,227,32,33,34,35]} endon
```
Note that there is no need for \ when entering the run manually

**Comment by Bacto** [comment]
Hi @barbudor,

Thanks for your reply. I really appreciate your help 🤩

I tried to create the rule by hand too with:
`Rule3 on file#mcp23x.dat do {"NAME":"My expander","GPIO":[224,225,226,227,32,33,34,35]} endon`

I enabled it via `Rule3 1` too, but I don't see any difference.

I didn't found documentation about `RuleX on file#...` so I activated UFS and created the `mcp23x.dat` file with that data too without success.

Also, I have no idea how `USE_MCP23XXX_DRV` will create the virtual switchs/buttons/relays. Will they appear as new GPIOs? On the main webpage? In Home Assistant?

As there is no documentation and I'm new to Tasmota I'm really having a hard time on this 😅

**Reply by barbudor** [reply]
You don't need to activate such rule because it's not really a rule. It just use the storage space dedicated for Rules as a pseudo file system

I haven't has the time to test the new driver yet but will try to test with your configuration tonight

**Comment by arendst** [comment]
It needs a restart to activate.

**Reply by Bacto** [reply]
I restarted but didn't saw anything new.

Just to be sure, can you tell me where to look to see if it worked?
What should I see new in the WEB UI?
Or what command should I run to see if it worked?

**Comment by arendst** [comment]
In your case with four relays you should see four buttons on the GUI with the relay state.

**Reply by Bacto** [reply]
Thanks for your help, I really appreciate :)


I restarted from scratch.

My configuration is the following :

- ESP8266EX (WEMOS D1 Mini)
- Pin out:
  - D1: I2C SCL
  - D2: I2C SDA
  - D3: WS2812
- MCP23008 expander (I2C) with:
  - GPIO 0 to 3: relays
  - GPIO 4 to 7: buttons


I compiled Tasmota v12.4.0 with only `USE_MCP23XXX_DRV` option (and no UFS):
```bash
echo '#define USE_MCP23XXX_DRV' >> tasmota/user_config_override.h
```

Configured the WiFi and connect to Tasmota web UI.

Then go to "Configure Other" put this and checked "Activate": `{"NAME":"Test board","GPIO":[1376,0,0,0,640,608,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1,"CMND":"rule3 on file#mcp23x.dat do {\"NAME\":\"Test expander\",\"GPIO\":[224,225,226,227,32,33,34,35]} endon"}`

After a restart, I see the WS2812 controls (that work) but nothing else.

<img width="437" alt="image" src="https://user-images.githubusercontent.com/1188497/223100112-edd293fa-c6a5-4419-bc26-28f1bf13c00f.png">


- `I2Scan` command returns `{"I2CScan":"Device(s) found at 0x20"}` which is the MCP23008 at address 0x20.
- `rule3` command returns `{"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":89,"Free":422,"Rules":"on file#mcp23x.dat do {\"NAME\":\"Test expander\",\"GPIO\":[224,225,226,227,32,33,34,35]} endon"}}`.

In web console with log level 4 ("more debug") I get this:
```
00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at FA, Count 18
00:00:00.055 QPC: Count 1
00:00:00.057 CFG: CR 356/699, Busy 0
00:00:00.071 ROT: Mode 1
00:00:00.073 SRC: Restart
00:00:00.077 Project tasmota - Tasmota Version 12.4.0(tasmota)-2_7_4_9(2023-03-06T11:15:00)
00:00:00.230 WIF: Checking connection...
00:00:00.231 WIF: Attempting connection...
00:00:00.567 WIF: Connecting to AP1 achgmbs Channel 6 BSSId <censored> in mode 11n as tasmota-<censored>...
00:00:01.752 WIF: Checking connection...
00:00:01.753 WIF: Attempting connection...
00:00:02.753 WIF: Checking connection...
00:00:02.754 WIF: Attempting connection...
00:00:03.757 WIF: Checking connection...
00:00:03.758 WIF: Attempting connection...
00:00:04.760 WIF: Checking connection...
00:00:04.761 WIF: Connected
00:00:05.015 HTP: Web server active on tasmota-<censored> with IP address 192.168.1.6
00:00:05.364 WIF: Sending Gratuitous ARP
00:00:05.366 NTP: Sync time...
00:00:05.392 WIF: DNS resolved '2.pool.ntp.org' (195.154.226.102) in 26 ms
00:00:05.416 RTC: UTC 2023-03-06T11:33:29, DST 2023-03-26T02:00:00, STD 2023-10-29T03:00:00
12:33:29.001 RTC: Synced by NTP
12:33:29.562 HTP: Main Menu
12:33:29.037 RSL: INFO1 = {"Info1":{"Module":"Test board","Version":"12.4.0(tasmota)","FallbackTopic":"cmnd/DVES_<censored>_fb/","GroupTopic":"cmnd/tasmotas/"}}
12:33:29.042 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-<censored>","IPAddress":"192.168.1.6"}}
12:33:29.052 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":8}}
12:33:29.060 RSL: RESULT = {"POWER":"OFF"}
12:33:29.064 RSL: POWER = OFF
12:33:29.411 QPC: Reset
12:33:30.744 HTP: Console
12:33:31.406 APP: Boot Count 8
12:33:31.562 CFG: Saved to flash at F9, Count 19, Bytes 4096
12:33:33.377 RSL: STATE = {"Time":"2023-03-06T12:33:33","Uptime":"0T00:00:11","UptimeSec":11,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Dimmer":10,"Color":"1A1A1A","HSBColor":"0,0,10","Channel":[10,10,10],"Scheme":0,"Width":1,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"achgmbs","BSSId":"<censored>","Channel":6,"Mode":"11n","RSSI":60,"Signal":-70,"LinkCount":1,"Downtime":"0T00:00:06"}}
12:33:38.416 RSL: config = {"ip":"192.168.1.6","dn":"Tasmota","fn":["Tasmota",null,null,null,null,null,null,null],"hn":"tasmota-<censored>","mac":"<censored>","md":"Test board","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"12.4.0","t":"tasmota_<censored>","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":1,"lt_st":3,"sho":[0,0,0,0],"sht":[[0,0,0],[0,0,0],[0,0,0],[0,0,0]],"ver":1}
12:33:38.420 RSL: sensors = {"sn":{"Time":"2023-03-06T12:33:38"},"ver":1}
```

Note: in "Configure Other", I don't see the "CMND" key but it seems ok as `rule3` seems to exist.

<img width="467" alt="image" src="https://user-images.githubusercontent.com/1188497/223100146-16014293-609c-4078-b1a1-02de829efc2b.png">


Do you have any idea or checks I can make?

**Reply by arendst** [reply]
Ah it's not yet in 12.4.0. You need Dev branch.

**Reply by Bacto** [reply]
Haaaa !
Thanks for that! I spent so much time on this 😅

So good news, it's now working with:
```json
{"NAME":"Bacto Multiusages Board","GPIO":[0,0,0,0,640,608,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1,"CMND":"rule3 on file#mcp23x.dat do {\"NAME\":\"Expander\",\"GPIO\":[224,225,226,227,64,65,66,67]} endon"}
```
<img width="470" alt="image" src="https://user-images.githubusercontent.com/1188497/223141101-5770d07d-eba1-4906-a9da-f80f3fa662df.png">


Nevertheless, if I add the WS2811 LED (1376) I don't have access to the buttons anymore, `POWER1` controls the LED and POWER2 to 5 don't work:
```json
{"NAME":"Bacto Multiusages Board","GPIO":[1376,0,0,0,640,608,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1,"CMND":"rule3 on file#mcp23x.dat do {\"NAME\":\"Expander\",\"GPIO\":[224,225,226,227,64,65,66,67]} endon"}
```
<img width="456" alt="image" src="https://user-images.githubusercontent.com/1188497/223141684-78cf21f6-7ecf-43da-9ecb-f02eacafe636.png">


Is it on purpose?

**Comment by arendst** [comment]
You might want to enable more logging to see what's going on.

**Reply by barbudor** [reply]
@arendst 
The problem is that the initialisation of the MCP23XXX driver is done in `XdrvCall(FUNC_MODULE_INIT)` 
But the loop in `XdrvCall()` will break `if (result && (function > FUNC_return_result))` 
When WS2812 is enabled, the Light driver (XDRV_04) will return True and will end the XDrvCall loop
So the MCT23XXX driver is never initialized

EDIT: This is a problem for when you are back. Hopefully you don't have a computer with you :roll_eyes:

**Reply by arendst** [reply]
Oops indeed. Was afraid some driver aborted the chain. Indeed. For next week.

**Reply by arendst** [reply]
Should be fixed in latest dev branch updated just now. Exepect pre-compiled binaries available in 2 hours.

**Reply by Bacto** [reply]
Hi,
Thanks for the update.
It now works! 🥰
